src/plugins/flow.js

```diff
@@ -1162,6 +1162,7 @@ export default function (instance) {
         let state = this.state.clone();
         try {
           let returnType = this.flowParseTypeAnnotation();
+          if (this.canInsertSemicolon()) this.unexpected();
```

Comment:
Thinking: we could just throw a different error, catch it, then throw with a nicer message instead:

```diff
 if (this.match(tt.colon)) {
   let state = this.state.clone();
   try {
     let returnType = this.flowParseTypeAnnotation();
-    if (this.canInsertSemicolon()) this.unexpected();
+    if (this.canInsertSemicolon()) throw new Error("IS_SEMICOLON_ERR");
     if (!this.match(tt.arrow)) this.unexpected();
     // assign after it is clear it is an arrow
     node.returnType = returnType;
   } catch (err) {
     if (err instanceof SyntaxError) {
       this.state = state;
+    } else if (err.message === "IS_SEMICOLON_ERR") {
+      this.unexpected(null, "Illegal newline before arrow");
     } else {
       throw err;
     }
   }
 }
```

It's not great, but it produces a much nicer error message, in the correct location too.

Thoughts?